### PR TITLE
MGMT-3884: Modified OCS validations to use all disks

### PR DESCRIPTION
**Pull request by priyanka19-98**
Each disk requires additional 2 cpus , 5gb ram
Signed off by: Priyanka Jiandani <pjiandan@redhat.com>

### CODE REVIEW

File: internal/operators/ocs/utils.go [hunk 1]
```diff
@@ -35,15 +35,16 @@ spec:
...
 type storageInfo struct {
-	OCSDisks int
+	OCSDisks int64
 }
 
 func OCSStorageCluster(StorageClusterManifest string, ocsDiskCounts int64) (string, error) {
 
 	// currently we are supporting only single storage set
 	// TODO support storage set according to disk count
+	ocsDiskCounts /= 3
```
Comment (avishayt) [inline internal/operators/ocs/utils.go:45]: The comment above is no longer relevant?

File: internal/operators/ocs/validations.go [hunk 2]
```diff
@@ -115,10 +117,21 @@ func (o *ocsValidator) ValidateOCSRequirements(cluster *models.Cluster) bool {
 			}
 		}
 	}
+
+	if DiskCount%3 != 0 {
+		status = "Total disks on the cluster must be a multiple of 3"
+		o.log.Info(status)
+		err := setOperatorStatus(cluster, status)
+		if err != nil {
+			o.log.Fatal("Failed to set Operator status ", err)
```
Comment (avishayt) [inline internal/operators/ocs/validations.go:126]: log.Fatal will cause the service to exit - we don't want that

Comment (avishayt) [reply internal/operators/ocs/validations.go:126]: I see that we have log.Fatal elsewhere in this file and in internal/ignition/ignition.go, please fix ASAP

File: internal/operators/ocs/validations.go [hunk 3]
```diff
@@ -156,16 +169,19 @@ func gbToBytes(gb int64) int64 {
...
 func (o *ocsValidator) validateOCS(log logrus.FieldLogger, hosts []*models.Host, cpu int64, RAM int64, disk int64, hostsWithDisks int64) (bool, string) {
-
+	var TotalCPUs int64
+	var TotalRAM int64
 	var status string
 	if int64(len(hosts)) == o.OCSRequiredHosts { // for 3 hosts
-		if cpu < o.OCSRequiredCompactModeCPUCount || RAM < gbToBytes(o.OCSRequiredCompactModeRAMGB) || disk < o.OCSRequiredDisk || hostsWithDisks < o.OCSRequiredHosts { // check for master nodes requirements
+		TotalCPUs = o.OCSRequiredCompactModeCPUCount + disk*o.OCSRequiredDiskCPUCount                                        // need additional 2 cpus per disk
+		TotalRAM = o.OCSRequiredCompactModeRAMGB + disk*o.OCSRequiredDiskRAMGB                                               // need additional 5 GB per disk
+		if cpu < TotalCPUs || RAM < gbToBytes(TotalRAM) || disk < o.OCSRequiredDisk || hostsWithDisks < o.OCSRequiredHosts { // check for master nodes requirements
```
Comment (avishayt) [inline internal/operators/ocs/validations.go:178]: This validation doesn't seem right to me - we should be checking that each host has the required resources for its disks, not looking at totals for the cluster (for example, we can have one host with many disks and little CPU, and another with few disks and lots of CPU).

Comment (priyanka19-98) [reply internal/operators/ocs/validations.go:193]: fixed

File: internal/operators/ocs/validations.go (hunk 2 above)
Comment (priyanka19-98) [reply internal/operators/ocs/validations.go]: fixed

File: internal/operators/ocs/utils.go (hunk 1 above)
Comment (priyanka19-98) [reply internal/operators/ocs/utils.go:44]: fixed

File: internal/operators/ocs/validations.go [hunk 4]
```diff
@@ -99,73 +103,104 @@ func (o *ocsValidator) ValidateOCSRequirements(cluster *models.Cluster) bool {
...
+	if DiskCount%3 != 0 {
+		status = "Total disks on the cluster must be a multiple of 3"
+		o.log.Info(status)
+		err := setOperatorStatus(cluster, status)
+		if err != nil {
+			o.log.Error("Failed to set Operator status ", err)
+		}
+		return false
+	}
 	// Hosts with disks must be a multiple of 3
```
Comment (nbalacha) [inline internal/operators/ocs/validations.go:134]: This comment does not seem to apply to the code immediately after it.

File: internal/operators/ocs/validations.go [hunk 5]
```diff
@@ -99,73 +103,104 @@ func (o *ocsValidator) ValidateOCSRequirements(cluster *models.Cluster) bool {
...
-		return err
+		return err, false
 	}
 
 	if len(disks) > 1 { // OCS must use the non-boot disks
+		diskCPU := int64((len(disks) - 1)) * o.OCSRequiredDiskCPUCount
+		diskRAM := int64((len(disks) - 1)) * o.OCSRequiredDiskRAMGB
+
+		if inventory.CPU.Count < diskCPU || inventory.Memory.UsableBytes < gbToBytes(diskRAM) {
+			status := fmt.Sprint("Resources are not sufficient on host with host ID ", host.ID, " to deploy OCS. A minimum of ", diskCPU, " CPUs, ", diskRAM, "RAMGB is required")
```
Comment (nbalacha) [inline internal/operators/ocs/validations.go:166]: Please reword to : "Insufficient Resources. Host <id> has <num> disks which requires a minimum of <diskCPU> cpu and <diskRAM>  memory"

or similar. It should include the number of disks on the host + the resources required.

File: internal/operators/ocs/validations.go [hunk 6]
```diff
@@ -99,73 +103,104 @@ func (o *ocsValidator) ValidateOCSRequirements(cluster *models.Cluster) bool {
...
+		status = "Total disks on the cluster must be a multiple of 3"
+		o.log.Info(status)
+		err := setOperatorStatus(cluster, status)
+		if err != nil {
+			o.log.Error("Failed to set Operator status ", err)
+		}
+		return false
+	}
 	// Hosts with disks must be a multiple of 3
 	o.OCSDisksAvailable = DiskCount
```
Comment (nbalacha) [inline internal/operators/ocs/validations.go:135]: This does not seem to be used anywhere.

File: internal/operators/ocs/validations.go [hunk 7]
```diff
@@ -69,24 +71,25 @@ func setOperatorStatus(cluster *models.Cluster, status string) error {
 func (o *ocsValidator) ValidateOCSRequirements(cluster *models.Cluster) bool {
 	var status string
 	hosts := cluster.Hosts
-
+	var err error
 	if int64(len(hosts)) < o.OCSRequiredHosts {
 		status = "Insufficient hosts to deploy OCS. A minimum of 3 hosts is required to deploy OCS. "
 		err := setOperatorStatus(cluster, status)
```
Comment (nbalacha) [inline internal/operators/ocs/validations.go:77]: this is a different err (not the one declared on line 74. Is that what was meant?

File: internal/operators/ocs/validations.go [hunk 8]
```diff
@@ -69,24 +71,25 @@ func setOperatorStatus(cluster *models.Cluster, status string) error {
...
-	var HostsWithDisks int64 = 0                 // to determine total number of hosts with disks. OCS requires atleast 3 hosts with non-bootable disks                           // to claim for the PV created by LocalVolumeSet
+	var cpuCount int64 = 0       //count the total CPUs on the cluster
+	var totalRAM int64 = 0       // count the total available RAM on the cluster
+	var diskCount int64 = 0      // count the total disks on all the hosts
+	var hostsWithDisks int64 = 0 // to determine total number of hosts with disks. OCS requires atleast 3 hosts with non-bootable disks
+	var insufficientHosts []string
 	if int64(len(hosts)) == o.OCSRequiredHosts { // for only 3 hosts, we need to install OCS in compact mode
 		for _, host := range hosts {
-			err := o.nodeResources(host, &CPUCount, &TotalRAM, &DiskCount, &HostsWithDisks)
+			err, insufficientHosts = o.nodeResources(cluster, host, &cpuCount, &totalRAM, &diskCount, &hostsWithDisks, insufficientHosts)
```
Comment (nbalacha) [inline internal/operators/ocs/validations.go:92]: Does this need to be returned if it is already being passed to the function as an argument?

File: internal/operators/ocs/validations.go [hunk 9]
```diff
@@ -99,108 +102,125 @@ func (o *ocsValidator) ValidateOCSRequirements(cluster *models.Cluster) bool {
...
+		status = "Total disks on the cluster must be a multiple of 3"
+		o.log.Info(status)
+		err := setOperatorStatus(cluster, status)
+		if err != nil {
+			o.log.Error("Failed to set Operator status ", err)
+		}
+		return false
+	}
+
+	// this wiil be used to set count of StorageDevices in StorageCluster manifest
```
Comment (nbalacha) [inline internal/operators/ocs/validations.go:146]: nit: typo in wiil

File: internal/operators/ocs/validations.go (hunk 7 above)
Comment (priyanka19-98) [reply internal/operators/ocs/validations.go]: fixed

File: internal/operators/ocs/validations.go (hunk 8 above)
Comment (priyanka19-98) [reply internal/operators/ocs/validations.go]: fixed

File: internal/operators/ocs/validations.go (hunk 9 above)
Comment (priyanka19-98) [reply internal/operators/ocs/validations.go]: fixed

**Review by avishayt — APPROVED** [review]

**Pull request merged**